crates/amaru-ledger/src/rules.rs

```diff
@@ -0,0 +1,193 @@
+mod block;
+
+use amaru_kernel::{alonzo::MaybeIndefArray, cbor, Block, Hash, Hasher, MintedBlock, Redeemers};
+
+use block::{
+    body_size::{block_body_size_valid, BlockBodySizeTooBig},
+    ex_units::*,
+    header_size::{block_header_size_valid, BlockHeaderSizeTooBig},
+};
+
+use crate::protocol_parameters::ProtocolParameters;
+
+pub enum RuleViolation {
+    BlockBodySizeTooBig(BlockBodySizeTooBig),
+    BlockHeaderSizeTooBig(BlockHeaderSizeTooBig),
+    TooManyExUnitsBlock(TooManyExUnits),
+}
+
+pub enum BlockValidationError {
+    SerializationError,
+    RuleViolations(Vec<RuleViolation>),
+    Composite(RuleViolation, Box<BlockValidationError>),
+}
+
+impl Into<BlockValidationError> for Vec<Option<RuleViolation>> {
+    fn into(self) -> BlockValidationError {
+        BlockValidationError::RuleViolations(
+            self.into_iter()
+                .filter(|r| r.is_some())
+                .map(|r| r.unwrap().into())
+                .collect(),
+        )
+    }
+}
+
+pub fn validate_block(
+    raw_block: &[u8],
+    protocol_params: ProtocolParameters,
+) -> Result<(Hash<32>, MintedBlock), BlockValidationError> {
+    let (block_header_hash, minted_block) = parse_block(raw_block)?;
+    let block: Block = minted_block.clone().into();
+
+    block_header_size_valid(minted_block.header.raw_cbor(), &protocol_params)
+        .map_err(|err| BlockValidationError::RuleViolations(vec![err.into()]))?;
+    block_body_size_valid(&minted_block, &protocol_params)
+        .map_err(|err| BlockValidationError::RuleViolations(vec![err.into()]))?;
```

Comment:
I still find this mapping into a singleton vector slightly awkward. Not a blocker but it just feels like something that is introduced _too early_. There's no need to collect multiple errors at this point.

Reply:
I understand this can be removed altogether by introducing a proper error hierarchy?

Reply:
I agree.